### PR TITLE
RSE-298: Case Category Permission Implementation.

### PR DESCRIPTION
**Pull request by tunbola**
## Overview
Since we have implemented the Case Category functionality, the case categories have been using the default case permissions. For example, the Manage Prospect, Prospect dashboard pages and functionality uses the same default permissions used for Manage Case and Case dashboard. 
There will be a need to separate these permissions and have permission set for each of the case category so that a user can have the prospect permission set and not have the case permission set.

Also there is a need to dynamically create this permission set for each case category. Once a case category is added, the permission set is dynamically added.

The permissions for a case category is created after the default case permissions. 
E.g The case category option name is used in the word replacement in below: 
```php
Default Case Permission: CiviCase: access my case and activities
Civi[casetype category option name]: access my [casetype category option name] and activities

For Prospect Category:
CiviProspect: access my prospect and activities

For Awards Category
CiviAwards: access my awards and activities
```
This is replicated for all the default case permissions.

## Before
- Case category permissions functionality not available.

## After
There are a couple of places where we need to implement dynamic case category permission checks. 
- The Case Create Form: `civicrm/case/add`: This page is used to create cases for any case category type, e.g cases, prospect by simply passing the `case_type_category` parameter in the URL. Civi checks permissions explicitly for this form ([See](https://github.com/civicrm/civicrm-core/blob/423616fa1bc75a6da428499e1680cfb8ec7c1c32/CRM/Case/Form/Case.php#L177)) 
- API calls from the frontend. Calls to Case.get and related API calls need to be checked so that the right permission will be used based on the case category being fetched for. Also delete, create API calls.

## Technical Details
**Approach**
- For the Case create form and other related PHP pages like the view case page that is used for all the case categories, the hook `hook_civicrm_permission_check` is used to check and grant permission based on the case type category under consideration
- For API calls, the hook `hook_civicrm_alterAPIPermissions` was used to alter the API permissions based on the case type category under consideration. For example when trying to delete a case e.g
```php
civicrm_api3(Case, delete, ['id' => 3]);
```
The case category that the case belongs to is checked, if it belongs to the prospect category, the the permission is set to the equivalent `delete prospect` permission. If the user performing the action does not have the permission, he is not able to delete.
> However for API calls, just modifying the permissions required to access them is not enough, there are also calls in the code that make permission checks directly when making API calls e.g when making a call to Case.get, permissions are checked [here](https://github.com/civicrm/civicrm-core/blob/master/CRM/Case/BAO/Case.php#L3099). To fix this issue, for FE requests the API calls are AJAX requests and we can easily pick the details of the API call and add an implementation in the `hook_civicrm_permission_check` to grant permission based on the case category.

- There are also instances where we need to check for case category permissions even when the API call is not for the Case entity. For example the `Activity.get` API makes permission checks based on the component the activity type belongs to ([See](https://github.com/civicrm/civicrm-core/blob/master/CRM/Activity/BAO/Activity.php#L95)). For these kind of instances, there is no way to know the case category under consideration, this was taken care of by ensuring that the user has at least any of the equivalent case category permissions.

### CODE REVIEW

**Comment by omarabuhussein** [comment]
Regarding this 

>>> There are also instances where we need to check for case category permissions even when the API call is not for the Case entity. For example the Activity.get API makes permission checks based on the component the activity type belongs to (See). For these kind of instances, there is no way to know the case category under consideration, this was taken care of by ensuring that the user has at least any of the equivalent case category permissions.

Cannot you get the case related to the activity from civicrm_case_activity table (or CaseActivity API endpoint) and from the case you get its category ?

File: CRM/Civicase/Hook/Permissions/CheckForActivityPageView.php [hunk 1]
```diff
@@ -1,69 +1,105 @@
...
+
+  /**
+   * Permission checks for activity page view.
+   *
+   * @param string $permission
+   *   Permission name.
+   * @param bool $granted
+   *   Whether permission is granted or not.
+   */
+  public function run($permission, &$granted) {
```
Comment (omarabuhussein) [inline CRM/Civicase/Hook/Permissions/CheckForActivityPageView.php]: I think only the run method need to be public

File: CRM/Civicase/Hook/Permissions/CheckForActivityPageView.php [hunk 2]
```diff
@@ -1,69 +1,105 @@
...
+    }
+  }
+
+  /**
+   * Checks request to see if page being accessed is an activity detail view.
+   *
+   * @return bool
+   *   Return Value.
+   */
+  public function isAccessingActivityPageView() {
```
Comment (omarabuhussein) [inline CRM/Civicase/Hook/Permissions/CheckForActivityPageView.php]: I am not sure how this method guarantee that we are on an activity view page

File: civicase.php [hunk 3]
```diff
@@ -498,10 +501,33 @@ function civicase_civicrm_postProcess($formName, &$form) {
  * Implements hook_civicrm_permission().
  */
 function civicase_civicrm_permission(&$permissions) {
-  $permissions['basic case information'] = [
-    'Civicase: basic case information',
-    ts('Allows a user to view only basic information of cases.'),
+  $permissionService = new CaseCategoryPermission();
+  // Add the permissions added by the civicase extension.
+  $caseCategoryPermissions = $permissionService->get();
```
[the remark this answers is not in the document]
Comment (tunbola) [reply civicase.php]: I get your point of view but the issue is the permission set defined for other case categories are already defined in core for Civicase, the only additional one added by our Civicase extension itself is the `BASIC_CASE_CATEGORY_INFO` permission, that's why the condition
```php
    if ($caseTypeCategory == CaseCategoryHelper::CASE_TYPE_CATEGORY_NAME) {
      continue;
    }
```
So as not to re-create those permissions.

File: CRM/Civicase/Hook/Permissions/CheckForActivityPageView.php (hunk 2 above)
Comment (tunbola) [reply CRM/Civicase/Hook/Permissions/CheckForActivityPageView.php]: tbh, not sure, maybe the method should have been renamed differently. This is not part of the changes for this PR, I only moved the file to a different location.

**Comment by tunbola** [comment]
on Comment
> Cannot you get the case related to the activity from civicrm_case_activity table (or CaseActivity API endpoint) and from the case you get its category ?

The call is from the front end and what is needed is actually the activities not the case bit when Civi detects the activity type is of that which belongs to a case the ACL code is triggered.

File: CRM/Civicase/Helper/CaseCategory.php [hunk 4]
```diff
@@ -20,16 +20,19 @@ class CRM_Civicase_Helper_CaseCategory {
...
-      'return' => ['case_type_id.case_type_category'],
-    ]);
+      if (!empty($result['case_type_id.case_type_category'])) {
+        $caseCategoryId = $result['case_type_id.case_type_category'];
 
-    if (!empty($result['case_type_id.case_type_category'])) {
-      $caseCategoryId = $result['case_type_id.case_type_category'];
+        return $caseTypeCategories[$caseCategoryId];
+      }
+    } catch (Exception $e) {
```
Comment (omarabuhussein) [inline CRM/Civicase/Helper/CaseCategory.php]: you should return null here

File: CRM/Civicase/Helper/CaseCategory.php [hunk 5]
```diff
@@ -47,13 +50,13 @@ public static function getCategoryName($caseId) {
   public static function getCategoryNameForCaseType($caseTypeId) {
     $caseTypeCategories = CaseType::buildOptions('case_type_category');
 
-    $result = civicrm_api3('CaseType', 'getsingle', [
+    $result = civicrm_api3('CaseType', 'getvalue', [
```
Comment (omarabuhussein) [inline CRM/Civicase/Helper/CaseCategory.php]: I think even getvalue throw an exception if no result found, so I think probably better to add Try, catch block

**Pull request merged**